### PR TITLE
change heating interface

### PR DESCRIPTION
**Pull request by jdannberg**
Change the interface of the heating model plugin so that it gets the material model inputs and outputs. 
This is a first step towards moving all of the heating terms into the heating model.

### CODE REVIEW

File: include/aspect/heating_model/interface.h [hunk 1]
```diff
@@ -79,6 +80,19 @@ namespace aspect
         update ();
 
         /**
+         * Function to compute the heating terms in @p heating_outputs given the
+         * inputs in @p material_model_inputs and the outputs of the material
+         * model in @p material_model_outputs.
+         * The default implementation calls specific_heating_rate to make
+         * this implementation backwards compatible.
```
Comment (tjhei) [inline include/aspect/heating_model/interface.h]: add a comment that says that heating_outputs is the value at each quadrature point as defined in inputs.

File: include/aspect/heating_model/interface.h [hunk 2]
```diff
@@ -79,6 +80,19 @@ namespace aspect
...
+         */
+        virtual
+        void
+        execute (const typename aspect::MaterialModel::Interface<dim>::MaterialModelInputs &material_model_inputs,
+                 const typename aspect::MaterialModel::Interface<dim>::MaterialModelOutputs &material_model_outputs,
+                 std::vector<double> &heating_outputs);
+
+        /**
          * Return the specific heating rate as a function of position.
```
Comment (tjhei) [inline include/aspect/heating_model/interface.h]: should we already deprecate ```specific_heating_rate```?

File: source/heating_model/interface.cc [hunk 3]
```diff
@@ -53,6 +53,22 @@ namespace aspect
 
     template <int dim>
     void
+    Interface<dim>::execute (const typename MaterialModel::Interface<dim>::MaterialModelInputs &material_model_inputs,
+                             const typename MaterialModel::Interface<dim>::MaterialModelOutputs &material_model_outputs,
+                             std::vector<double> &heating_outputs)
+    {
+      for (unsigned int i=0; i<heating_outputs.size(); ++i)
```
Comment (tjhei) [inline source/heating_model/interface.cc]: Assert size of heating_outputs and inputs.position is equal?

**Comment by tjhei** [comment]
looks good otherwise.

File: source/postprocess/internal_heating_statistics.cc [hunk 4]
```diff
@@ -89,15 +89,17 @@ namespace aspect
 
             this->get_material_model().evaluate(in, out);
 
+            std::vector<double> heating_model_outputs(n_q_points);
```
Comment (bangerth) [inline source/postprocess/internal_heating_statistics.cc]: Please move the declaration of this array out of the loop.

File: include/aspect/simulator.h [hunk 5]
```diff
@@ -612,6 +612,7 @@ namespace aspect
       double compute_heating_term(const internal::Assembly::Scratch::AdvectionSystem<dim>  &scratch,
                                   typename MaterialModel::Interface<dim>::MaterialModelInputs &material_model_inputs,
                                   typename MaterialModel::Interface<dim>::MaterialModelOutputs &material_model_outputs,
+                                  const double heating_model_output,
```
Comment (bangerth) [inline include/aspect/simulator.h]: I can't figure out what an *input* argument called `...output` does. Can you rename or document?

Comment (bangerth) [inline include/aspect/simulator.h]: I see from the code below. How about name it `specific_heating_rate`?

File: include/aspect/heating_model/interface.h [hunk 6]
```diff
@@ -79,6 +80,21 @@ namespace aspect
...
+         * filled with the value of the heating rate at each quadrature point as
+         * defined in @p material_model_inputs.
+         * The default implementation calls specific_heating_rate to make
+         * this implementation backwards compatible.
+         */
+        virtual
+        void
+        execute (const typename aspect::MaterialModel::Interface<dim>::MaterialModelInputs &material_model_inputs,
+                 const typename aspect::MaterialModel::Interface<dim>::MaterialModelOutputs &material_model_outputs,
+                 std::vector<double> &heating_outputs) const;
```
Comment (bangerth) [inline include/aspect/heating_model/interface.h]: A function called `execute` typically does some action, but in common use of the word does not return anything. How about renaming it to `compute_specific_heating_rates` or similar?

**Comment by bangerth** [comment]
Looks pretty good -- please address the issues I just commented on, add an entry to changes.h and merge yourself.

File: source/heating_model/interface.cc [hunk 7]
```diff
@@ -53,6 +53,25 @@ namespace aspect
 
     template <int dim>
     void
+    Interface<dim>::evaluate (const typename MaterialModel::Interface<dim>::MaterialModelInputs &material_model_inputs,
+                              const typename MaterialModel::Interface<dim>::MaterialModelOutputs &material_model_outputs,
+                              std::vector<double> &heating_outputs) const
+    {
+      for (unsigned int i=0; i<heating_outputs.size(); ++i)
+        {
+          Assert(heating_outputs.size() == material_model_inputs.position.size(),
```
Comment (tjhei) [inline source/heating_model/interface.cc]: move outside the loop

File: source/heating_model/interface.cc [hunk 8]
```diff
@@ -53,6 +53,25 @@ namespace aspect
 
     template <int dim>
     void
+    Interface<dim>::evaluate (const typename MaterialModel::Interface<dim>::MaterialModelInputs &material_model_inputs,
+                              const typename MaterialModel::Interface<dim>::MaterialModelOutputs &material_model_outputs,
+                              std::vector<double> &heating_outputs) const
+    {
+      for (unsigned int i=0; i<heating_outputs.size(); ++i)
```
Comment (tjhei) [inline source/heating_model/interface.cc]: q is a more appropriate name for the loop variable

**Pull request merged**